include/seqan3/alignment/matrix/detail/coordinate_matrix.hpp

```diff
@@ -0,0 +1,233 @@
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::coordinate_matrix.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/std/ranges>
+
+#include <seqan3/alignment/matrix/detail/matrix_coordinate.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief A function object that converts a column index and a row index to a seqan3::detail::matrix_coordinate.
+ * \ingroup alignment_matrix
+ *
+ * \details
+ *
+ * This helper function object is used for the seqan3::detail::coordinate_matrix_iterator to convert a column index
+ * and row index into a seqan3::detail::matrix_coordinate. The seqan3::detail::coordinate_matrix_iterator iterates
+ * over the columns of the underlying coordinate matrix and returns a transformed iota range over the row indices of
+ * this column. The column index is stored as member inside of this function object since every column index is
+ * associated with many row indices.
+ */
+struct convert_to_matrix_coordinate
+{
+    //!\brief The index of the represented column.
+    size_t column_index{0};
+
+    /*!\brief The conversion operator
+     *
+     * \param[in] row_index The index of the represented row.
+     *
+     * \returns seqan3::detail::matrix_coordinate for the current column and row index.
+     */
+    auto operator()(size_t const row_index) noexcept
+    {
+        return matrix_coordinate{row_index_type{row_index}, column_index_type{column_index}};
+    }
+};
+
+/*!\brief The iterator for the seqan3::detail::coordinate_matrix.
+ * \ingroup alignment_matrix
+ * \implements std::forward_iterator
+ *
+ * \details
+ *
+ * Iterates over the columns of the underlying seqan3::detail::coordinate_matrix. The iterator returns a transformed
+ * std::ranges::views::iota view over the row indices. In the transformation every row index is converted to a
+ * seqan3::detail::matrix_coordinate using the seqan3::detail::convert_to_matrix_coordinate function object.
+ */
+class coordinate_matrix_iterator
+{
+private:
+
+    //!\brief The currently represented column index.
+    size_t column_id{0};
+    //!\brief The size of a column.
+    size_t column_size{0};
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief The value type.
+    using value_type = decltype(std::views::iota(size_t{0u}, size_t{1u})
+                              | std::views::transform(convert_to_matrix_coordinate{column_id}));
+    //!\brief The reference type.
+    using reference = value_type;
+    //!\brief The pointer type.
+    using pointer = void;
+    //!\brief The difference type.
+    using difference_type = std::ptrdiff_t;
+    //!\brief The iterator category.
+    using iterator_category = std::forward_iterator_tag;
```

Comment:
AFAIK you would only need iterator concept if you want to interface with some pre-c++20 library because contiguous Tag was first introduced then. This allows you to conform with the new iterator concepts while still providing a legacy iterator.